### PR TITLE
Adding a Pull Request Template

### PR DESCRIPTION
**Pull request by a-smith-github**
# GitHub Infrastructure: PR Template
## Fixes \#297

# Change Description

This change adds a single markdown file to the hidden .github folder which will be used by GitHub to pre-populate the pull request form when this change reaches the default branch (main). I therefore propose this branch merges directly into main.

Some previous pull request descriptions have been short or relying on similarly brief issue descriptions. To aid the review process, this change aims to encourage developers to note the motivations behind each change as well as to outline their design and what limitations a reviewer might need to be aware of to make an informed review. This should hopefully speed up review, improve review quality and potentially enhance traceability/reduce engineering effort in the future. 

Please feel free to provide feedback / directly suggest changes during your review.

This change will close Issue #297 

<br>

# Test Description

I've not added a test for this change. I've not (yet) tried running lint/style analysis manually on this. 

<br>

# Documentation Impact

This changes development documentation at pull request time and references the README. The README informed the checklist at the bottom of the document. Synchronisation may need to be performed manually as changes are made.

<br>
<hr>

### Pre-Request Checklist

- [ x ] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [ x ] No new warnings are generated
- [  N/A ] The documentation has been updated (or an issue has been created to track the corresponding change)
- [  N/A ] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [ x  ] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [ N/A ] File dates have been updated to reflect modification date
- [  ] This change conforms to the conventions described in the README

### CODE REVIEW

**Comment by mondus** [comment]
Please add me as a reviewer for this.

**Comment by a-smith-github** [comment]
I'll investigate this failure which appears to occur outside of this change

**Comment by mondus** [comment]
@a-smith-github Not a review comment but a general comment. Your markdown syntax in the PR description is slightly incorrect for the checklist as they are not rendering as checkable boxes.

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+
+<br>
+<hr>
+
+### Pre-Request Checklist
+
+- [  ] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
```
Comment (mondus) [inline .github/pull_request_template.md:70]: MD syntax is incorrect. It should be ` - [ ]`.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+<br>
+
+# Change Description
+<em>
+Please write a summary of the change and specify which Issue this fixes. This should cover the motivation for making the change and provide almost enough context to review the change in isolation. While the issue description is a good place to discuss the failure/requirements, this is the ideal space for explaining why this design fixes the issue and why you designed it the way you did.
+
+<br>
+
+- If there are important design decisions, particularly those that leverage functionality outside this change or were caused by existing limitations, they should be noted here.
+- If there is a design specification which relates to this PR, please provide a link to it here.
```
Comment (mondus) [inline .github/pull_request_template.md:13]:
```suggestion
- If there is a design specification which relates to this PR, please provide a link to it here. Alternatively if there is an issue then link the issue with the PR by using the closes #issue_number syntax.
```

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# Pull Request Title
+## Fixes \#(your issue number)
+
+<br>
```
Comment (mondus) [inline .github/pull_request_template.md:4]: Suggest to remove HTML syntax and retain pure markdown

Comment (mondus) [reply .github/pull_request_template.md:4]: There are a number of such places to do this.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+Please write a summary of the change and specify which Issue this fixes. This should cover the motivation for making the change and provide almost enough context to review the change in isolation. While the issue description is a good place to discuss the failure/requirements, this is the ideal space for explaining why this design fixes the issue and why you designed it the way you did.
+
+<br>
+
+- If there are important design decisions, particularly those that leverage functionality outside this change or were caused by existing limitations, they should be noted here.
+- If there is a design specification which relates to this PR, please provide a link to it here.
+- If there are any dependencies (third-party or internal downstream) which are added or impacted, please note them here. And if their limitations informed your design, please note how they did so here (particularly if this information was not captured in a design specification).
+</em>
+
+Text goes here
```
Comment (mondus) [inline .github/pull_request_template.md:17]: I would suggest to encourage in the template a list of checklist items. E.g.

It is useful to provide a [task list](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists) within the pull request. These should breakdown the overall PR into units of work which quite often can be associated to either a specific issue (if the PR closes multiple) or a commit. The task list allows the developer to update on development progress while the PR is in draft. It provides insight into the current state of the PR at any specific point in time. Once the PR is complete (ready for review) it is expected that each item would be checked or justification provided as to why the task will not be completed in this PR (perhaps as it has moved to a new issue). An example task list:

 - [ ] Define a test
 - [ ] Implement working version
 - [ ] Document feature by providing worked example
 - [ ] Update README

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+<br>
+
+# Test Description
+<em>
+Please describe the tests you have written that verify the enhancement works or that the bug fix has been remedied. For a bug fix, please ensure a complete description of the failure is available here or in the issue such that a reviewer can verify that the testing is suitable. This is a good place to highlight any manual testing you may have had to perform. 
+
+Please describe any operating system limitations placed on the testing (if appropriate).
+
+If your tests fix bugs, please explain that in the comment you provided in the test file and please record the mapping here e.g.
```
Comment (mondus) [inline .github/pull_request_template.md:30]: Not limited to bugs this is also to describe closing of features if defined within issues.

**Review by mondus — CHANGES_REQUESTED** [review]
Looks good but I have made some suggestions. I think the most important is promoting the use of tasks lists in the PR description. I have provided some text for this that you could include. There are a few more minor syntax and style changes.

**Comment by einola** [comment]
This looks good @a-smith-github - and even better with Paul's changes.

I understand that GitHub will use this template for PRs into main, but what about PRs into develop - can we make GitHub use it there as well?

**Comment by draenog** [comment]
> This looks good @a-smith-github - and even better with Paul's changes.
> 
> I understand that GitHub will use this template for PRs into main, but what about PRs into develop - can we make GitHub use it there as well?

As far as I know, GitHub always uses the  template stored in a  default branch (in our case main) for all pull requests, independent which branch they target. So this change should affect also PRs into develop.

File: .github/pull_request_template.md (hunk 3 above)
Comment (a-smith-github) [reply .github/pull_request_template.md:4]: Happy to swap but <br /> or adding a &nbsp; is the only thing that works in my editor for adding spaces between headings. Appreciate doing something like this is outside CommonMark (the github one) --- is this bad practice anyway? Google is ambiguous, I'd love to get your view. For now I'll drop the extra line break and remove html tags

File: .github/pull_request_template.md (hunk 2 above)
Comment (a-smith-github) [reply .github/pull_request_template.md:13]: Accepting with change

```suggestion
- If there is a design specification which relates to this PR, please provide a link to it here.
- If the PR relates to an open issue issue, please link the issue to the PR by using the 'closes #issue_number' syntax.
```

File: .github/pull_request_template.md (hunk 5 above)
Comment (a-smith-github) [reply .github/pull_request_template.md:31]: Agree, expanding text

File: .github/pull_request_template.md (hunk 4 above)
Comment (a-smith-github) [reply .github/pull_request_template.md:21]: I have a pre-PR checklist but agree this task this serves a special purpose. Given the ability to provide at-a-glance development transparency, I'm going to put this at the very top under the 'Fixes Issue Number' line

**Comment by einola** [comment]
> I _think_ the local link to the design spec will work correctly when it's in main but I'd appreciate if someone else with better knowledge (or an example repository to compare against) could comment. Alternatively (perhaps sub-optimally) could fix later if becomes a bug (non-breaking) - it'll be easier - or could use full url (if I'm mistaken with this attempt, I suspect that this is the correct solution)

I'd support merging and fixing later if there's a problem. It's not crucial if we introduce a non-breaking bug into main at this point. But let's wait a bit and see if there's a counter-point to this.

> The plan right now is to wait on merges into main to resolve the test failure which depends on a file that no longer exists into develop. Instead of handling the merge conflict, the general view is to wait until develop merges into main. Would happily take broader input on this strategy too.

We did that already and all the checks pass - so that doesn't prevent the merge anymore.

**Comment by a-smith-github** [comment]
Thanks Einar --- I hadn't noticed the break had been fixed. I've pushed a new change, if that is also successful then I think we can proceed.

I've figured out a better way of testing the local url and found a few use-cases where global url could be an issue (prevents local clones from functioning correctly) so despite being a fix for this issue, it would have created a limitation. On further examination, my understanding of how the relative path worked on GitHub was incorrect however after testing I have modified the relative path to be correct - although it should be reviewed. I don't think this will cause a doc-only bug anymore but if it does, as we've both noted, will be non-breaking and easy to fix.

**Review by mondus — APPROVED** [review]

**Comment by dorchard** [comment]
Great- this has been approved now, but "At least 1 approving review is required by reviewers with write access.". @einola do you want to do this, or @draenog can?

**Comment by einola** [comment]
I was surprised that Paul didn't have write access, but I've fixed this now (by giving the nextsimdg developers team write access).

**Pull request merged**